File: Kuznetsov_Ilja/2039.py
from typing import List

class Solution:
    # @staticmethod
    def networkBecomesIdle(self, edges: List[List[int]], patience: List[int]) -> int:
        # building graph
        vertices_num = len(patience)
        neighbors: List[List[int]] = [[] for _ in range(vertices_num)]
        for from_idx, to_idx in edges:
            neighbors[from_idx].append(to_idx)
            neighbors[to_idx].append(from_idx)
        
        # BFS
        current_step_indices = [0]
        next_step_indices: List[int] = []
        visited = [False] * vertices_num
        visited[0] = True
        
        cur_time = 1
        max_idle_moment = -1
        
        while current_step_indices:
            cur_vertex_idx = current_step_indices.pop()
            for neighbor_idx in neighbors[cur_vertex_idx]:
                if not visited[neighbor_idx]:
                    next_step_indices.append(neighbor_idx)
                    visited[neighbor_idx] = True
                    
                    pat = patience[neighbor_idx]
                    max_idle_moment = max(
                        max_idle_moment,
                        4 * cur_time - (2 * cur_time - 1) % pat
                    )
            
            if not current_step_indices:
                current_step_indices = next_step_indices
                next_step_indices = []
                cur_time += 1
        
        return max_idle_moment

File: 2039.cpp
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    int networkBecomesIdle(vector<vector<int>>& edges, vector<int>& patience) 
    {
        int n = patience.size();
        vector<vector<int>> graph(n);
        for (auto &i : edges)
        {
            graph[i[0]].emplace_back(i[1]);
            graph[i[1]].emplace_back(i[0]);
        }

        vector<int> path(n, INT_MAX);
        path[0] = 0;
        queue<int> q;
        q.push(0);
        while(!q.empty())
        {
            int nd = q.front();
            q.pop();
            for (int i = 0; i < graph[nd].size(); ++i)
            {
                if (path[graph[nd][i]] == INT_MAX)
                {
                    path[graph[nd][i]] = path[nd] + 1;
                    q.push(graph[nd][i]);
                }
            }
        }
        int ans = 0;
        for (int i = 1; i < n; ++i)
        {
            int path_size = path[i] * 2;
            int p = path_size % patience[i];

            if (path_size <= patience[i]) ans = max(ans, path_size + 1);
            else if (p != 0) ans = max(ans, path_size + path_size - p + 1);
            else ans = max(ans, path_size + path_size - patience[i] + 1);
        }
        return ans;
    }
};